server/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java

```diff
@@ -600,4 +560,40 @@ public void onFailure(Exception e) {
             }
         });
     }
+
+    private static class TestListener implements ShardStateAction.Listener {
+
+        private final SetOnce<Exception> failure = new SetOnce<>();
+        private final CountDownLatch latch = new CountDownLatch(1);
+
+        @Override
+        public void onSuccess() {
+            try {
+                failure.set(null);
+            } finally {
+                latch.countDown();
+            }
+        }
+
+        @Override
+        public void onFailure(final Exception e) {
+            try {
+                failure.set(e);
+            } finally {
+                latch.countDown();
+            }
+        }
+
+        boolean isSuccessful() {
+            return getFailure() == null;
+        }
+
+        Exception getFailure() {
+            return failure.get();
+        }
+
+        void await() throws InterruptedException {
```

Comment:
I guess you could argue, that it leaves less wiggle room for mistakes if it's just one method.
The way it was and still is, I have to wait for and then check the result. But if I check before I waited, I simply get the wrong result without any indication that I'm doing something wrong.
=> It's a bit of a test instability waiting to happen for some the failure without waiting :D
+ I guess you could also just name the method `result` or `get` and it would be pretty clear/ergonomic like a `Future`.

But no big deal imo :)